### PR TITLE
Update library to use definitions of ultrahdr_api everywhere

### PR DESCRIPTION
**Pull request by ram-mohan**
This is a major change but maintains bitexactness with previous commit. This change unifies legacy structure definitions with definitions of ultrahdr_api.h. This helps for better extensibility for new features and avoid redundancy.

Legacy structures are moved to ultrahdr.h. These are deprecated and only retained for backward compatibility.

Briefly,
- unify ultrahdr_color_gamut with uhdr_color_gamut_t
- unify ultrahdr_transfer_function with uhdr_color_transfer_t
- unify ultrahdr_metadata_struct with uhdr_gainmap_metadata_t
- unify jpegr_uncompressed_struct with uhdr_raw_image_t
- unify jpegr_compressed_struct with uhdr_compressed_image_t
- unify jpegr_exif_struct with uhdr_mem_block_t
- unify status_t with uhdr_error_info_t
- Deprecate ultrahdr_output_format
- Added methods to Jpeg*Helper to simplify data translation between helper
  and its users

Bug fix,
- For images with multi channel gainmap, decoded gainmap is not copied
  completely for getter functions. This is fixed.
- Add support for tonemapping linear transfer inputs
- Fixes oss-fuzz: 69287

Test: ./ultrahdr_unit_test
Test: ./ultrahdr_dec_fuzzer
Test: ./ultrahdr_enc_fuzzer

### CODE REVIEW

File: lib/include/ultrahdr/ultrahdr.h [hunk 1]
```diff
@@ -111,6 +116,69 @@ struct ultrahdr_metadata_struct {
...
+  // Pixel format.
+  uhdr_img_fmt_t pixelFormat = UHDR_IMG_FMT_UNSPECIFIED;
+  // Color range.
+  uhdr_color_range_t colorRange = UHDR_CR_UNSPECIFIED;
+};
+
+/*
+ * Holds information for compressed image or gain map.
+ */
+struct jpegr_compressed_struct {
```
Comment (DichenZhang1) [inline lib/include/ultrahdr/ultrahdr.h:158]: Should this also changed to "uhdr_*"

**Comment by DichenZhang1** [comment]
Thank you for the change, but we may need to hold on for a little bit. jpegr.h layer is currently used by Android, and this change will also need change in Android and it's out of scope now. We will hold this change for a while and land it with v2 version.

**Comment by DichenZhang1** [comment]
I would recommend separate this change land the bug fixing part now, and hold the API change part. Let me know what you think. Thank you!

File: lib/include/ultrahdr/ultrahdr.h (hunk 1 above)
Comment (ram-mohan) [reply lib/include/ultrahdr/ultrahdr.h:158]: This is not required. I have kept the legacy type names as is with out making any changes to maintain compatibility with android. Earlier, the uhdr_compressed_image_t is converted to jpegr_compressed_struct and sent to decodeJPEGR. Now the reverse happens. Through out the library this change is now using uhdr_* structs and the legacy structures are mapped to uhdr_* structs.

For example,
encode API-0
  uhdr_error_info_t encodeJPEGR(uhdr_raw_image_t* hdr_intent, uhdr_compressed_image_t* dest,
                                int quality, uhdr_mem_block_t* exif) { ... } 
is the current implementation.

 // legacy implementation of encode API - 0
  status_t encodeJPEGR(jr_uncompressed_ptr p010_image_ptr, ultrahdr_transfer_function hdr_tf,
                       jr_compressed_ptr dest, int quality, jr_exif_ptr exif) { ... } is the alias of the above. 

This function maps its arguments to uhdr_* and calls current implementation.

To summarize, ultrahdr.h contains all the legacy structures and are marked for deprecation.

**Comment by ram-mohan** [comment]
> Thank you for the change, but we may need to hold on for a little bit. jpegr.h layer is currently used by Android, and this change will also need change in Android and it's out of scope now. We will hold this change for a while and land it with v2 version.

This change does not effect android builds. The android builds works normally.

**Comment by DichenZhang1** [comment]
OK Thanks, I'll take a further look

**Comment by DichenZhang1** [comment]
> > Thank you for the change, but we may need to hold on for a little bit. jpegr.h layer is currently used by Android, and this change will also need change in Android and it's out of scope now. We will hold this change for a while and land it with v2 version.
> 
> This change does not effect android builds. The android builds works normally. It seems recent sample app changes are causing build issues in android. These were corrected in this commit.
> 
> > I would recommend separate this change land the bug fixing part now, and hold the API change part. Let me know what you think. Thank you!
> 
> Some fixes work with the changes made. I will try to seperate independent one's and give a pull request.

OK then we are having two interfaces of the public APIs. I'm not against with this move, but I'm feeling this is not an urgent need. And additionally this change is fairly too large and I would still recommend separating this change and we'll merge the bug fixing part first. Let me know what you think. thank you!

**Comment by ram-mohan** [comment]
> OK then we are having two interfaces of the public APIs. 

As the interface is from ultrahdr_api.h, there is only one. But if we were to include jpegr.h directly then as you have rightly pointed out, there are aliases for each api.

> I'm not against with this move, but I'm feeling this is not an urgent need. And additionally this change is fairly too large and I would still recommend separating this change and we'll merge the bug fixing part first. 

Certain fixes that are independent of this change, i have moved them out and issued pull requests. #178 #179 #180 
But some fixes where a little more coupled with this change (oss-fuzz 69287 and returning decoded gainmap). So i did not move them. Also, some parts of this change simplifies handling 444 and 422 support. 

> When I use the attached sources, I get error "encountered unknown error during decoding" (exit code 255). 
I assume this is related to the JPG, though it does not say which file decoding failed (would be nice to update the error message to be more clear there).

It also addresses this.

Some portions of this change are not at all a priority and can be marked as good to have and some other parts are helpful towards addressing known issues.

**Comment by DichenZhang1** [comment]
OK my only concern of this change is about having aliases at jpegr.cpp layer, but indeed this change makes the code unified and clean, and has benefit of simplifying 444 and 422 color format. I've approved this change. Thank you for work!

**Review by DichenZhang1 — APPROVED** [review]

File: examples/ultrahdr_app.cpp [hunk 2]
```diff
@@ -298,6 +296,10 @@ class UltraHdrAppInput {
         mQuality(95),
         mOTf(oTf),
         mOfmt(oFmt),
+        mFullRange(UHDR_CR_UNSPECIFIED),
+        mMapDimensionScaleFactor(1),
```
Comment (DichenZhang1) [inline examples/ultrahdr_app.cpp:300]: Same comment as I mentioned in the other CL

**Review by DichenZhang1 — APPROVED** [review]

File: examples/ultrahdr_app.cpp (hunk 2 above)
Comment (ram-mohan) [reply examples/ultrahdr_app.cpp]: Done

Comment (ram-mohan) [reply examples/ultrahdr_app.cpp]: Done

**Review by DichenZhang1 — APPROVED** [review]
